### PR TITLE
Seminar 5 materials

### PR DESCRIPTION
**Pull request by VladVin**

### CODE REVIEW

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+   "source": [
+    "То же самое сделаем для точек из пространства малой размерности:  \n",
+    "$$q_{j|i}=\\frac{exp(-\\lVert{y_i - y_j}\\rVert^2)}{\\sum_{k!=i}{exp(-\\lVert{y_i - y_k}\\rVert^2)}}$$"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "На самом деле это выражение взято из оригинальной версии SNE алгоритма. В алгоритме t-SNE вместо нормального распределения используется t-распределение Стьюдента со степению свободы 1 ([Wikipedia](https://en.wikipedia.org/wiki/Student%27s_t-distribution)). Эта модификация сделана для ускорения процесса сходимости, о котором мы скажем позже. Итак,\n",
```
Comment (strawberrypie) [inline Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Может как-то переписать чтобы было понятнее, без углубления в детали? (Кажется упоминание распределения Стьюдента и степеней свободы лишние)

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+   "source": [
+    "Мы получили две функции измерения похожести объектов в двух различных пространствах (!). Т.е., если наша модель идеально предсталяет точки в новом пространстве, то $p_{j|i}$ будет равен $q_{j|i}$."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Поскольку точки $y_i$ - это те точки, которые мы собираемся сгенерировать, нам нужно генерировать их по определенному закону. Введем расстояние Кульбака-Лейблера для измерения расстояния между $P_i$ и $Q_i$, где $P_i$ - условное распределение вероятности всех точек $x_j$ кроме $j!=i$, имея $x_i$, а $Q_i$ - условное распределение вероятности всех точек $y_j$ кроме $j!=i$, имея $y_i$:  \n",
+    "$$C = \\sum_{i}{KL({P_i}\\lVert{Q_i})} = \\sum_{i}{\\sum_{j}{p_{j|i}log\\frac{p_{j|i}}{q_{j|i}}}}$$"
```
Comment (strawberrypie) [inline Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Что такое `C`? Насколько я понял, это функция потерь, которую мы пытаемся минимизировать. Упомяни это.

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+   "metadata": {},
+   "source": [
+    "Теперь для нахождения оптимального решения (точек $y_i$) мы можем использовать метод градиентного спуска. После схождения метода, полученные точки можно отрисовать в нашем \"человеко-читаемом\" пространстве. Результат работы алгоритма t-SNE на датасете MNIST:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![t-SNE for MNIST](data/t_sne_mnist.png)"
```
Comment (strawberrypie) [inline Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Добавь ссылку на http://projector.tensorflow.org/
Там поиграться можно интерактивно с MNIST и с другими датасетами.
Кажется визуализацией можно прекрасно завершить эту часть.

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+   "metadata": {},
+   "source": [
+    "### Визуализация с помощью matplotlib"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Тут будет немного знакомства с визуализацией matplotlib"
```
[the remark this answers is not in the document]
Comment (VladVin) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Может не будет, и я удалю это

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+   "metadata": {},
+   "source": [
+    "Разные классы объектов довольно сильно пересекаются. Если мы уберем цветовую раскраску, то отличить некоторые классы на этой картинке будет просто невозможно."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Переформулируем немного задачу:** нам необходимо разместить точки из пр-ва высокой размерности в пр-ве низкой размерности так, чтобы близкие объекты в исходном пр-ве были близки в новом, и наоборот, далекие объекты в исходном пр-ве были так же далеки в новом."
```
[the remark this answers is not in the document]
Comment (VladVin) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Чтобы меньше места занимать текстом. Ну ладно, уже двое сказали, что лучше полностью писать, испралю

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Чтобы измерить похожесть объектов, введем условную вероятность присутствия в определенном месте одной точки при условии присутствия другой:  \n",
+    "$$p_{j|i}=\\frac{exp(-\\lVert{x_i - x_j}\\rVert^2 / 2\\sigma_i^2)}{\\sum_{k!=i}{exp(-\\lVert{x_i - x_k}\\rVert^2/2\\sigma_i^2)}}$$"
```
[the remark this answers is not in the document]
Comment (VladVin) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Просто картинка это один слайд у авторов, на следующем они говорят, что на самом деле можно использовать разные сигмы для каждого объекта.
Насчет равно спасибо, исправлю

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb (hunk 1 above)
Comment (VladVin) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Ну тогда будет просто формула, кот. без описания просто вставлена и все. Опять же - кому знает, те поймут. Остальные просто проигнорируют

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb (hunk 2 above)
Comment (VladVin) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Можно, да

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb (hunk 1 above)
Comment (lemhell) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: "На самом деле" здесь лишнее.

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb (hunk 3 above)
Comment (VladVin) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: А я сейчас как раз вставляю sklearn код для MNIST и еще какой-нибудь датасет с лицами

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb (hunk 2 above)
Comment (lemhell) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Если это функция потерь, то лучше записать ее как `L`, как обычно пишут.
Если у тебя используются какие-то символы и они не пояснены, то пояснять их не "можно", а нужно.

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (strawberrypie) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Проблема в том, что используется символ и нигде не описано, что он значит.
Не понятно, почему в распределении иксов она есть, а в игреках нет.

File: Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb (hunk 2 above)
Comment (strawberrypie) [reply Seminar 5/Seminar_5.2_t-SNE_visualization.ipynb]: Кстати, нашел ответ про отличия loss function, cost function и objective function: https://stats.stackexchange.com/questions/179026/objective-function-cost-function-loss-function-are-they-the-same-thing
